register Node/UpdateTaint event to plugins which has Node/Add only and doesn't have Node/UpdateTaint

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

register a failure plugin from preCheck into pInfo.UnschedulablePlugins so that we could technically prevent https://github.com/kubernetes/kubernetes/issues/109437 happening again.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #122284
Fixes #122306
Part of #122305

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Currently, NodeAdded QueueingHint could not always be called because of the internal feature called preCheck.
It's definitely not something expected for plugin developers,
and we're trying to eventually remove preCheck completely to fix this.
Until then we'll register UpdateNodeTaint event for plugins that have NodeAdded event, but don't have UpdateNodeTaint event.
It'd result in a bad impact on the requeuing efficiency though, a lot better than some Pods being stuck in the
unschedulable pod pool.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

sanposhiho committed:
register Node/UpdateNodeTaint event to plugins which has Node/Add only, doesn't have Node/UpdateNodeTaint

sanposhiho commented:
@kerthcet This is my proposal. I know I have to add UTs but open it now to show you my idea. 
What do you think?

AxeZhan commented on pkg/scheduler/scheduler.go:
```diff
@@ -385,17 +385,45 @@ func buildQueueingHintMap(es []framework.EnqueueExtensions) internalqueue.Queuei
 			}
 
+			if event.Event.Resource == framework.Node {
+				if event.Event.ActionType&framework.Add == 0 {
```
Shouldn't this be `event.Event.ActionType&framework.Add != 0` ?

sanposhiho replied:
Yes! 😅

kerthcet commented:
We have two approaches:
- Option one is what Kensei did here
- Option two is add the missing update event to plugins like DRA as we did to other plugins as well

And I have two concerns:
- Do we need to cherry pick this fix to 1.28, 1.29, I think so because this will make some pods unschedulable unhonouring the node events.
- Do we need to consider out-of-tree plugins, if so, Kensei's fix is a more general way, or we should only consider the in-tree plugins, then option2 seems more simple.

kerthcet commented:
cc @kubernetes/sig-scheduling-misc

sanposhiho commented:
Thanks, that's a good summary. I'm on the same page with @kerthcet. 

> Do we need to cherry pick this fix to 1.28, 1.29, I think so because this will make some pods unschedulable unhonouring the node events.

+1

> Do we need to consider out-of-tree plugins, if so, Kensei's fix is a more general way, or we should only consider the in-tree plugins, then option2 seems more simple.

I think Yes, so I'd prefer this approach basically.
Also, I think we have another option which is to take both approaches - we cherry-pick option2 one because that's gonna be a simple and enough change to fix this issue in the default scheduler. And, we also merge this PR (option1) for the mitigation for custom plugins, but not cherry-pick.

sanposhiho commented:
Given no clear objection to this idea so far, I'll work on making it ready this weekend.

alculquicondor commented:
Can we do this only if the featuregate for scheduling hints is enabled?

sanposhiho commented:
Regardless of whether QHint is enabled or not, this bug could happen when EventsToRegister has NodeAdd, but doesn't have NodeUpdate. (https://github.com/kubernetes/kubernetes/issues/109437)

sanposhiho commented:
/retitle register Node/Update event to plugins which has Node/Add only, doesn't have Node/Update

Now it's ready to get reviews.

sanposhiho commented:
/priority important-soon

sanposhiho commented:
I'll check the failing tests tomorrow.

sanposhiho commented:
/retest

sanposhiho commented:
/retest

alculquicondor commented:
Shouldn't this change be guarded by the feature gate?
Otherwise we might be doing unnecessary retries when not using scheduling hints

sanposhiho commented:
@alculquicondor 

This change should be applied no matter whether the feature gate is enabled or not.

Thinking about this scenario:
1. Node is created but marked as unready. 
2. (1) arrives in the scheduler via EventHandler, but preCheck drops this event because of un-ready.
3. Node is now ready. 
4. (3) arrives in the scheduler via EventHandler, preCheck says OK this time.

For example, PodTopologySpread has NodeAdd and UpdateNodeLabel in EventsToHandler (before this PR). 
PodTopologySpread wants to requeue Pods at NodeAdd (2), but (2) doesn't arrive in the scheduling queue because of preCheck. And, because PodTopologySpread doesn't have NodeUpdate, (4) doesn't move Pods rejected by PodTopologySpread. This problematic scenario could happen regardless of the feature gate value.

alculquicondor commented:
OTOH, the existing functionality has been in production for a while, with no bug reports. So I wonder how commonly this happens?

Two requests:
1. Add integration tests to proof that this was a problem (specially around the NodeReady scenario).
2. Maybe we should consider targeting updates to NodeReady, as opposed to any update. Targeting any update has the potential of causing more regressions due to increased retries.

kerthcet commented:
I'm inclined to be conservative here, this is a change to all the plugins, both in-tree and out-of-tree. We should be cautious. Maybe change the in-tree plugins one by one(one or two plugins we have) is enough unless we got bug reports. 🥲

sanposhiho commented:
> Add integration tests to proof that this was a problem (specially around the NodeReady scenario).

We already have it:
https://github.com/kubernetes/kubernetes/blob/c0a1ded053e51a448c2e96d7c476fa19ae781711/test/integration/scheduler/queue_test.go#L180

It only checks if noderesourcefit encounters this bug. 
It's better to have per-plugin enqueueing tests to cover all plugins though, it's gonna be huge (already tracked in https://github.com/kubernetes/kubernetes/issues/122305)

> Maybe we should consider targeting updates to NodeReady, as opposed to any update. Targeting any update has the potential of causing more regressions due to increased retries.

Can you elaborate? Adding a new ActionType like [UpdateNodeLabel, etc](https://github.com/kubernetes/kubernetes/blob/v1.28.0/pkg/scheduler/framework/types.go#L52-L55)?

alculquicondor commented:
If we already have the test, then it's not a problem?

Add the remaining integration tests in separate PRs.

> Can you elaborate? Adding a new ActionType like [UpdateNodeLabel, etc](https://github.com/kubernetes/kubernetes/blob/v1.28.0/pkg/scheduler/framework/types.go#L52-L55)?

yes

> I'm inclined to be conservative here, this is a change to all the plugins, both in-tree and out-of-tree. We should be cautious.

Even if we just "fix" in-tree plugins, we might not see any problems until we test at scale (5k nodes, each with 100 pods, for example).

bart0sh commented:
/remove-sig node

kerthcet commented:
Take the DRA for example, I mean update `UpdateNodeLabel` to `UpdateNode`  for safe fix and solve the `preCheck` trouble in this release, then we can revert to `UpdateNodeLabel`.

sanposhiho commented:
> If we already have the test, then it's not a problem?

That test proves only that it's not a problem in NodeResourceFit. For other plugins that are changed in this PR, this problem comes in.

>> Can you elaborate? Adding a new ActionType like [UpdateNodeLabel, etc](https://github.com/kubernetes/kubernetes/blob/v1.28.0/pkg/scheduler/framework/types.go#L52-L55)?
>
>yes

Got it, agree, I think that'll be a good initial step for the mitigation since we expect Node's readiness change is the major trigger of this issue. We can think of other triggers as well (by other factors implemented in `preCheck`) though, those shouldn't be as major as Node readiness and we can delay the fix until we get actual bug reports.
Let me change this PR.

sanposhiho commented:
Updated this PR not to add Node/Update but add Node/UpdateNodeTaint.
I initially thought we could have a new NodeUpdate event to target the updates for `node.kubernetes.io/not-ready` taint like @alculquicondor suggested, but I ended up using UpdateNodeTaint because I saw another NoSchedule taint `node.cloudprovider.kubernetes.io/uninitialized` was added to the Node being initialized in my personal cluster (GKE).

```
$ kubectl get nodes -w
...
gke-cluster-1-default-pool-b2e6ce80-8nq5   NotReady   <none>   1s      v1.27.3-gke.100


$ kubectl get nodes gke-cluster-1-default-pool-b2e6ce80-8nq5 -o=yaml
...
  spec:
    taints:
    - effect: NoSchedule
      key: node.cloudprovider.kubernetes.io/uninitialized
      value: "true"
    - effect: NoSchedule
      key: node.kubernetes.io/not-ready
```

(Of course, if everyone wants, we can still create another NodeUpdate event to target the updates for `node.kubernetes.io/not-ready` and `node.cloudprovider.kubernetes.io/uninitialized` though.)

---

Also, I improved the TestCoreResourceEnqueue test to show this bug existing. https://github.com/kubernetes/kubernetes/pull/122538/ is another PR to make sure this test fails if we haven't got this PR's changes.

sanposhiho commented:
/retitle register Node/UpdateTaint event to plugins which has Node/Add only and doesn't have Node/UpdateTaint

alculquicondor commented:
Can you update the PR description with the chosen approach?

alculquicondor commented on pkg/scheduler/framework/types.go:
```diff
@@ -71,7 +71,16 @@ const (
+	// NodeAdd QueueingHint isn't always called because of the internal feature called preCheck.
+	// It's definitely not something expected for plugin developers,
+	// and registering NodeUpdated event is the only mitigation for now.
+	// So, we'll register NodeUpdated event for plugins that has NodeAdded event, but doesn't have NodeUpdated event.
```
```suggestion
	// So, kube-scheduler registers UpdateNodeTaint event for plugins that has NodeAdded event, but don't have NodeUpdated event.
```

alculquicondor commented on pkg/scheduler/scheduler.go:
```diff
@@ -385,17 +385,47 @@ func buildQueueingHintMap(es []framework.EnqueueExtensions) internalqueue.Queuei
+			// It's definitely not something expected for plugin developers,
+			// and registering NodeUpdated event is the only mitigation for now.
+			//
+			// So, here, we'll register NodeUpdated event for plugins that has NodeAdded event, but doesn't have NodeUpdated event.
```
this is out of date.

alculquicondor commented on pkg/scheduler/scheduler.go:
```diff
@@ -385,17 +385,47 @@ func buildQueueingHintMap(es []framework.EnqueueExtensions) internalqueue.Queuei
 				fn = defaultQueueingHintFn
 			}
 
+			if event.Event.Resource == framework.Node {
```
Can we just add the action here?

sanposhiho replied:
We have to go through all events to judge whether we should add the UpdateNodeTaint event or not because users can register multiple events for the same resource in EventsToRegister like this:

```go
func (pl *CustomPlugin) EventsToRegister() []framework.ClusterEventWithHint {
	return []framework.ClusterEventWithHint{
		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterNodeAdd},
		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.UpdateNodeTaint}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
	}
}
```

sanposhiho commented on test/integration/scheduler/queue_test.go:
```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
 // TestCoreResourceEnqueue verify Pods failed by in-tree default plugins can be
 // moved properly upon their registered events.
 func TestCoreResourceEnqueue(t *testing.T) {
```
Made it table-driven. It'll be the basis for https://github.com/kubernetes/kubernetes/issues/122305.

sanposhiho commented:
@alculquicondor Please take a look again.

sanposhiho commented:
/retest

sanposhiho commented:
@kubernetes/sig-scheduling-approvers Can anyone take a look?

alculquicondor commented:
after enhancements freeze? I have too much to review these days.

alculquicondor commented on test/integration/util/util.go:
```diff
@@ -1157,6 +1157,8 @@ func NextPodOrDie(t *testing.T, testCtx *TestContext) *schedulerframework.Queued
 }
 
 // NextPod returns the next Pod in the scheduler queue, with a 5 seconds timeout.
+// Note that this function leaks goroutines in the case of timeout; even after this function returns after timeout,
```
Let's open an issue for this. It should be possible to send a Broadcast to the sync.Cond in the queue.

sanposhiho replied:
https://github.com/kubernetes/kubernetes/issues/123480

alculquicondor replied:
Can you revert this so that we don't depend on an owner for this file?

pohly replied:
/approve

As test owner.

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
+// scheduleNextPod pops the next Pod from the scheduling queue and schedule it manually.
+// It returns the name of the Pod that was popped.
+func scheduleNextPod(t *testing.T, testCtx *testutils.TestContext, scheduleSuccess bool) {
 	podInfo := testutils.NextPodOrDie(t, testCtx)
```
call `t.Helper`?

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
-	// Now we should be able to pop the Pod from activeQ again.
+// scheduleNextPod pops the next Pod from the scheduling queue and schedule it manually.
+// It returns the name of the Pod that was popped.
+func scheduleNextPod(t *testing.T, testCtx *testutils.TestContext, scheduleSuccess bool) {
```
This has some repeated production code... Maybe we can call `schedulingCycle` instead?

sanposhiho replied:
I changed the test to use `ScheduleOne()` instead of `scheduleNextPod`. (`schedulingCycle()` could only replace a few lines in `scheduleNextPod` since we'd still have to pop a Pod from the queue. )

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
+					t.Fatalf("Failed to trigger the event: %v", err)
+				}
+
+				time.Sleep(1 * time.Second)
```
why the sleep?

sanposhiho replied:
Changed it to wait.PollUntilContextTimeout.

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
+		triggerFn func(testCtx *testutils.TestContext) error
+		// requeuedPods is the list of Pods that are expected to be requeued after triggerFn.
+		// The order matters - the first Pod in the list is expected to be requeued first after triggerFn.
+		requeuedPods sets.Set[string]
```
```suggestion
		wantRequeuedPods sets.Set[string]
```

alculquicondor commented on test/integration/scheduler/queue_test.go:
```diff
@@ -178,89 +179,185 @@ func TestSchedulingGates(t *testing.T) {
+					}); err != nil {
+						t.Fatal(err)
+					}
+					scheduleNextPod(t, testCtx, true /* this Pod is supposed to be scheduled */)
```
Maybe just create the Pod with a node name?

ahg-g commented on pkg/scheduler/scheduler.go:
```diff
@@ -385,17 +385,47 @@ func buildQueueingHintMap(es []framework.EnqueueExtensions) internalqueue.Queuei
+			// This behavior will be removed when we remove the preCheck feature.
+			// See: https://github.com/kubernetes/kubernetes/issues/110175
+			queueingHintMap[framework.ClusterEvent{Resource: framework.Node, ActionType: framework.UpdateNodeTaint}] =
+				append(queueingHintMap[framework.ClusterEvent{Resource: framework.Node, ActionType: framework.UpdateNodeTaint}],
```
Why is this specific to taints? what about a node that receives a node label update that impacts the nodeAffinity preCheck? Also, why doesn't the plugin explicitly add those instead of implicitly doing it here (which seems to me the right thing to do)?

sanposhiho replied:
cc @alculquicondor 

> what about a node that receives a node label update that impacts the nodeAffinity preCheck

https://github.com/kubernetes/kubernetes/pull/122292#issuecomment-1862867359 changed this PR's direction from adding a general node update to adding a node taint event.
You're actually right. If we want a complete fix (covering all minor scenarios), we should register `Node/Update` event here, but it'd hurt the throughput by requeueing many Pods unnecessarily.

So, the point would be which scenario(s) is likely enough to deserve a fix. And, I believe [this scenario](https://github.com/kubernetes/kubernetes/pull/122292#issuecomment-1861794637) is, while others aren't.

ahg-g replied:
Ok, I see. I think the right solution is to add the taint/toleration plugin to the list of UnschedulablePlugins for that pod when it fails preCheck on node add event. We could change the function signature of PreCheck to return the plugin that failed the pod.

ahg-g replied:
We are discussing the "right" solution for this PR. Since PreChecks continue to exist, then having them integrate with failedPlugins is certainly the better approach than this hack which could be surprising to plugins and is not really covering all corner cases.

I didn't get a chance to look at the implementation, but I didn't expect it to be that complicated.

sanposhiho commented on pkg/scheduler/framework/plugins/podtopologyspread/plugin.go:
```diff
@@ -148,7 +148,7 @@ func (pl *PodTopologySpread) EventsToRegister() []framework.ClusterEventWithHint
 		// Node add|delete|update maybe lead an topology key changed,
 		// and make these pod in scheduling schedulable or unschedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Delete | framework.Update}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.Delete | framework.UpdateNodeLabel | framework.UpdateNodeTaint}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
```
It should be OK. Actually before https://github.com/kubernetes/kubernetes/pull/122195, we only has `UpdateNodeLabel` rather than `Update`. (was changed to `Update` to deal with [this issue](https://github.com/kubernetes/kubernetes/issues/122284))

sanposhiho commented on pkg/scheduler/framework/types.go:
```diff
@@ -71,7 +71,16 @@ const (
-	Pod                   GVK = "Pod"
+	Pod GVK = "Pod"
+	// There is a note about Node Add event and UpdateNodeTaint event.
+	// NodeAdd QueueingHint isn't always called because of the internal feature called preCheck.
+	// It's definitely not something expected for plugin developers,
+	// and registering UpdateNodeTaint event is the only mitigation for now.
+	// So, kube-scheduler registers UpdateNodeTaint event for plugins that has NodeAdded event, but don't have UpdateNodeTaint event.
+	// It has a bad impact for the requeuing efficiency though, a lot better than some Pods being stuck in the
+	// unschedulable pod pool.
+	// This behavior will be removed when we remove the preCheck feature.
+	// See: https://github.com/kubernetes/kubernetes/issues/110175
```
Just for the visibility. The plugin developers would (hopefully) read this comment when they're using `Node` event. If we only write this comment in pkg/scheduler/scheduler.go, it'd be hidden behaviour for them.

sanposhiho commented:
@ahg-g addressed your suggestion https://github.com/kubernetes/kubernetes/pull/122292#discussion_r1501895093.

/retitle register a failure plugin from preCheck into `pInfo.UnschedulablePlugins` for a correct requeueing

kerthcet commented:
I **respect** all the efforts @sanposhiho made here, but my poor points is still we'd better not to make the code more complicated based on the assumption that we'll remove precheck one day. The alternative is follow what we did before by simply adding more events in registration. 

My two cents, I think you can make better decision on this 👉 cc @ahg-g @sanposhiho

alculquicondor commented:
I agree that it might be wiser not to complicate the precheck logic that is intended for removal in the near future.

sanposhiho commented:
@kerthcet @alculquicondor 
Agree with the complexity. Reverted the last commit anyways as the code freeze is very close.

/retitle register Node/UpdateTaint event to plugins which has Node/Add only and doesn't have Node/UpdateTaint

alculquicondor commented:
/retest

alculquicondor reviewed (commented):
just some comments are off.

/approve

ahg-g commented:
/hold

alculquicondor commented:
I didn't have a chance to review the new commit, so I'm more comfortable with what I had already reviewed before. But I'll defer to you, if you feel confident about the alternate approach.

ahg-g commented:
> What is the complexity? Looking [c967331](https://github.com/kubernetes/kubernetes/commit/c967331ecedc017e9cee74b8809fa15ff0ea53b6) , the implementation seems quite reasonable.

If this logic is needed, then I agree it is complicated and better not to go that route. But why is it needed? Why not just insert the unschedulable plugin to `pInfo.UnschedulablePlugins` if preCheck fails here https://github.com/kubernetes/kubernetes/blob/4369fcb00fc7c5bd623ddcc95d6d1d46092c804d/pkg/scheduler/internal/queue/scheduling_queue.go#L1108?

sanposhiho commented:
@ahg-g 

> Why not just insert the unschedulable plugin to pInfo.UnschedulablePlugins if preCheck fails here

Two reasons in my mind when I implemented c967331ecedc017e9cee74b8809fa15ff0ea53b6:
- We want to insert the plugin name from preCheck **only when** `event` is interesting for some `pInfo.UnschedulablePlugins`.
  - Otherwise, it'd result in a regression. For example, every time an unready node event comes, all Pods would get `TaintToleration` in `pInfo.UnschedulablePlugins`, even though some of them may be waiting for only Pod events to arrive.
- `Pending` status handling; we should run preCheck after Pending plugins are executed.
  - Otherwise, if some of `pInfo.PendingPlugins` are waiting for some Node events but preCheck filters Node events out, it might result in Pods being stuck forever at unschedQ.
  - (Note that, this is a problem for the current master branch too, so basically, it aims to solve an additional existing problem.)

If we want to do it in moveAllToActiveOrBackoffQueue, it'd be like

```go
func (p *PriorityQueue) moveAllToActiveOrBackoffQueue(logger klog.Logger, event framework.ClusterEvent, oldObj, newObj interface{}, preCheck PreEnqueueCheck) {
	if !p.isEventOfInterest(logger, event) {
		// No plugin is interested in this event.
		// Return early before iterating all pods in unschedulablePods for preCheck.
		return
	}
	
	unschedulablePods := make([]*framework.QueuedPodInfo, 0, len(p.unschedulablePods.podInfoMap))
	for _, pInfo := range p.unschedulablePods.podInfoMap {
		if !interestToUnschedulablePlugins(pInfo, event) { continue } // This Pod isn't interested in this event, ignore.

		if preCheck != nil {
			if admissionResult := preCheck(pInfo.Pod); admissionResult != nil {
				if admissionResult.Name != "" {
					// If a plugin rejects a Pod in preCheck, we should treat the pod as rejected by the plugin.
					// Here, we may register non-registered plugins to pInfo.UnschedulablePlugins if a user uses a custom scheduler configuration though,
					// we don't have to care about it because it's just ignored.
					pInfo.UnschedulablePlugins.Insert(admissionResult.Name)
				}
				continue // This event is filtered out by preCheck, ignore.
			}
		}

		unschedulablePods = append(unschedulablePods, pInfo) 
	}
	p.movePodsToActiveOrBackoffQueue(logger, unschedulablePods, event, oldObj, newObj)
}

func interestToUnschedulablePlugins(pInfo, event) bool {
	rejectorPlugins := pInfo.UnschedulablePlugins.Union(pInfo.PendingPlugins)

	for eventToMatch, hintfns := range p.queueingHintMap[pInfo.Pod.Spec.SchedulerName] {
		if !eventToMatch.Match(event) {
			continue
		}

		for _, hintfn := range hintfns {
			if !rejectorPlugins.Has(hintfn.PluginName) {
				// skip if it's not hintfn from rejectorPlugins.
				continue
			}
			return true // some pInfo.UnschedulablePlugins are interested in this event.
		}
	}
	return false // no pInfo.UnschedulablePlugins are interested in this event.
}

```

But, I prefer the original way https://github.com/kubernetes/kubernetes/commit/c967331ecedc017e9cee74b8809fa15ff0ea53b6 asc967331ecedc017e9cee74b8809fa15ff0ea53b6 addresses Pending plugin issue as well and require no additional `for` loop.

---

> What is the complexity? 

Actually, I prefer c967331ecedc017e9cee74b8809fa15ff0ea53b6 to the current way of implementing it (register `node/UpdateTaint`). But, I admit that it'd increase complexity, which is not in the implementation itself, but in the fact that it increases the scenarios that we have to consider around `pInfo.UnschedulablePlugins`. Until now, the only scenario is it's coming from the scheduling cycle, but after c967331ecedc017e9cee74b8809fa15ff0ea53b6, it may be added by preCheck.

ahg-g commented:
Thanks @sanposhiho for the explanation. The logic in https://github.com/kubernetes/kubernetes/pull/122292#issuecomment-1979894913 is certainly a lot simpler than https://github.com/kubernetes/kubernetes/commit/c967331ecedc017e9cee74b8809fa15ff0ea53b6 and it is more inline with what I thought we could do.

I will leave it to you to decide which approach to go with, it seems both have their own drawbacks.

/hold cancel

ahg-g commented:
/retest

sanposhiho commented:
@ahg-g I changed the implementation for https://github.com/kubernetes/kubernetes/pull/122292#issuecomment-1979894913 's direction then as I don't have a strong opinion; 
- As noted in the comment, the issue with PreCheck/Pending plugin is a separate issue, and I believe no in-tree plugin is impacted by this bug for now anyway. 
- Then I'd rather take the readability of the implementation for everyone by having additional `for`, which would be a big deal for overall efficiency.

@kerthcet @alculquicondor I'd like to proceed this way, if you think it's now acceptable complexity-wise.

sanposhiho commented:
/retitle register a failure plugin from preCheck into `pInfo.UnschedulablePlugins` for a correct requeuein

Sorry for moving back and forth 😅

ahg-g commented:
I would stick with the original one, it had more eyes on it, so lets just proceed with that.

sanposhiho commented:
Rebased and squashed into one commit with fixing [one point from Aldo last time](https://github.com/kubernetes/kubernetes/pull/122292#discussion_r1513080638).

sanposhiho commented:
/retitle register Node/UpdateTaint event to plugins which has Node/Add only and doesn't have Node/UpdateTaint

alculquicondor commented:
/lgtm
/approve
Let me discuss with the leads whether this qualifies for test freeze.

alculquicondor commented:
Also, it might be worth changing plugins such as NodeResourcesFit to use UpdateTaint instead of Update. But we can leave that for after the code freeze.

sanposhiho commented:
@MrHohn Can you take a look at this PR? Only minor comment change is your area.

alculquicondor commented:
@pohly could you give us a hand?

alculquicondor commented:
@ahg-g and I agreed that, as a bug, this qualifies for the test freeze

/milestone v1.30

pohly commented:
> could you give us a hand?

I looked through the discussion thread, but I am too much out of the topic to have a specific opinion.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -396,7 +396,7 @@ func (pl *dynamicResources) EventsToRegister() []framework.ClusterEventWithHint
 		// A resource might depend on node labels for topology filtering.
 		// A new or updated node may make pods schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel | framework.UpdateNodeTaint}},
```
Isn't the code behind `registerNodeAdded && !registerNodeTaintUpdated` below adding this automatically? Does it still make sense to also add this manually?

A short comment here that this is a workaround would be useful to avoid confusion when reading just this file, because nothing in it depends on node taints.

sanposhiho replied:
> Does it still make sense to also add this manually?

Yes, especially for QHint - When you add QHint for Node/Add, you probably want to have the same QHint for Node/UpdateNodeTaint, which you couldn't do if we let the scheduler automatically add `UpdateNodeTaint`.

> A short comment here that this is a workaround would be useful to avoid confusion when reading just this file, because nothing in it depends on node taints.

👍

pohly replied:
> you probably want to have the same QHint for Node/UpdateNodeTaint,

This is the part that I don't get. This particular plugin doesn't check taints of a node. It leaves that to the other plugins which check whether a pod fits the node.

If you are adding it because "it might be needed", then it might be necessary. If it's getting added because "node added" is not delivered reliably (= i.e. it's a workaround), then it makes more sense.

alculquicondor replied:
It's not that "node added" events are not delivered reliably.

It's that a new node is likely not ready (it has a not-ready taint). And it needs to be retried when it becomes ready.

sanposhiho commented:
@pohly Your approval would be helpful, as you're a [test/OWNERS](https://github.com/kubernetes/kubernetes/blob/master/test/OWNERS) approver. We only have a small comment update in test/ area.
